### DISCUSSION TITLE
Logging - different log levels for log handlers

### DISCUSSION THREAD

**Discussion by mabartos** [quarkusio/quarkus#43042] [Q&A] [answered]
Hey Quarkus community and team!

Is there any way, how can I configure different log levels for categories and different log handlers?

I mean, I'd like to set the logger level differently for log handlers.

Example:

- Console -  `."org.keycloak".level=INFO`
- File - `."org.keycloak".level=DEBUG`
- Syslog - `."org.keycloak".level=TRACE`

And simultaneously still respect the root level for other loggers (`quarkus.log.level`, `quarkus.log.console.level`,...)

It might be doable by leveraging filters, but I'm wondering, whether there's a better approach. 

cc: @dmlloyd

**Comment by dmlloyd** [comment]
Normally you would set the category level to the most detailed level that you want to appear in *any* log. Then you'd set the level on each handler to the most detailed level that you want to see for that particular destination. So e.g. your console log would globally have a level of `INFO`, while the `org.keycloak` category would be configured for `TRACE`.

There isn't a built-in category×handler level table. The reason is that each category and each handler can store the target level as an integer, and can filter the messages using a fast integer comparison (which is important, especially for `TRACE` logging which can be quite verbose). But a combinatorial table would add some significant complexity.

For this reason, the best approach is to start as above by setting your category levels first, and then your handler levels. Then, if there are still certain messages which are leaking through which you don't want, applying a filter to the handler which filters down certain categories (with e.g. a string switch) would be the next logical step.

If you end up in a situation where you have a very large number of these exceptional cases to consider, we could potentially revisit this as some kind of feature enhancement where individual category configurations could specify levels per handler.

**Comment by mabartos** [comment]
@dmlloyd Thanks for your comment.

> For this reason, the best approach is to start as above by setting your category levels first, and then your handler levels. Then, if there are still certain messages which are leaking through which you don't want, applying a filter to the handler which filters down certain categories (with e.g. a string switch) would be the next logical step.

I'm not sure if this is the way we want to go, as it might become quite fragile, and also the maintainability would not be very good. Especially handling these exceptional cases.

For Keycloak, we'd like to have the possibility to manage the categories for each individual log handler separately to provide more fine-grained log level settings. 

It means that, for instance, for console, we'd like to have `info` root level and `debug` level for `org.keycloak.events` (or whatever) category, but for file log handler having `debug` root level and `trace` level for `org.keycloak.events`. And I think it's not feasible to do it with the mentioned suggestion.

> If you end up in a situation where you have a very large number of these exceptional cases to consider, we could potentially revisit this as some kind of feature enhancement where individual category configurations could specify levels per handler.

IMHO, it might be very useful to have the ability to specify log level for category per log handler and it would be great to revisit and create an RFE.

--------------------------------------------
I've been trying to provide such functionality with the help of Filters in Keycloak, and it works fine. I made only a brief performance analysis for handling `trace` records, and in terms of CPU samples, there is no major difference. 

Here is the PR: https://github.com/keycloak/keycloak/pull/32620/files#diff-fb502b5da4f070cc324f0ad9a7fc495e1b06e5fafc4979dd1ed17a079dc7cf7aR34.

It would be great if you could provide some feedback on the used filters as you have more expertise than me.

Thank you very much!

**Comment by dmlloyd** [comment]
I think the idea is basically sound. I would suggest that you ensure that `foo.bar` is not considered a parent category to `foo.barbaz`. But I think this is a pretty good solution.

**Comment by mabartos** [comment]
@dmlloyd Can we transfer this discussion to RFE? It'd be great for us to have such capabilities included in Quarkus instead of making some workarounds in Keycloak. By transferring it into the RFE, we could see whether somebody else has the same requirements and if it's wanted by the community. WDYT?

Do you think the category×handler level table would introduce any big major performance degradation?

**Reply by dmlloyd** [reply]
Sounds like a good idea!

**Comment by mabartos** [comment, marked as answer]
Moved to RFE issue: https://github.com/quarkusio/quarkus/issues/43147
